frontend/model/database.js

```diff
@@ -9,12 +9,42 @@ const log = localforage.createInstance({
   storeName: 'Contracts'
 })
 
+const isLightweightClient = process.env.LIGHTWEIGHT_CLIENT === 'true'
+
 // make gi.log use localforage for storage
-sbp('sbp/selectors/overwrite', {
-  'gi.db/get': key => log.getItem(key),
-  // TODO: handle QuotaExceededError
-  'gi.db/set': (key, value) => log.setItem(key, value)
-})
+sbp('sbp/selectors/overwrite',
+  isLightweightClient
+    ? {
+        'gi.db/get': key => {
+          const contractId = sbp('gi.db/log/contractIDFromLogHEAD', key)
+          if (!contractId) {
+            return null
+          }
+          const state = sbp('state/vuex/state')
+          if (state?.contracts?.[contractId]?.HEAD) {
```

Comment:
Are the `?` checks on `state?` and `state.contracts?` necessary? I don't see how they could be null unless something went really wrong.

Reply:
Perhaps not, although I'm not sure what the alternative would be to not doing the checks (a runtime exception?). I've replaced this to `if (state.contractsXXX[contractId].HEADXX) {` for testing, and, as expected, no compile time errors occur. I guess that I've grown used to being a bit defensive in JS when it comes to these things. I can remove the null / undefined checks if you think a hard fail is better.

Reply:
@snowteamer What are your thoughts on this? Should I just remove the `?`s?

Reply:
I had no idea that we could do this now. Is this a feature of `esbuild`? Because last I remember this required a special babel plugin...

I'm personally OK with it, but I would remove some of the `?`s that you can know for sure don't need to be there. For example, checking `state?` isn't necessary, that is guaranteed to be there by the time this function is called. Each `?` leads to a branch, and the more branches you add to the code, the more things slow down (it adds up eventually).

Similarly, I think it's safe to remove the `?` for `.contracts?`, because if `state` is defined, that should be defined too as it's initialized with it.

Finally... maybe a hard fail is better, since by adding these `?`s I count the following downsides:

- Slightly slower code
- Loss of awareness of bugs
- Potential breakage if we switch bundlers (e.g. to Parcel #1107)

And only the following upside:

- `?.` is fantastic in situations where you *know* that values can be non-existent, and want to avoid uglifying your code with massive if/else blocks

Since this situation doesn't fit that — we know that all of these values should exist — I vote for removing them to catch bugs.

Reply:
Yea I agree with greg here